dungeon/src/contrib/utils/DynamicCompiler.java

```diff
@@ -0,0 +1,165 @@
+    }
+
+    // Load compiled class using custom loader to override any previous versions
+    MyClassLoader classLoader = new MyClassLoader(new URL[] {outputRoot.toURI().toURL()});
+    Class<?> loadedClass = classLoader.loadClass(className);
+    sourceFile.delete();
+
+    return loadedClass;
+  }
+
+  /**
+   * Dynamically instantiates a class from a source file using an optional set of constructor
+   * arguments.
+   *
+   * <p>This method compiles the provided Java source file, loads the resulting class using a custom
+   * class loader, and creates a new instance using the specified constructor arguments.
+   *
+   * @param sourcePath Path to the Java source file.
+   * @param className Fully qualified name of the class to load (including package name). This must
+   *     exactly match the package and class declaration inside the file.
+   * @param args Optional constructor arguments as {@code Tuple<Class<?>, Object>}: each tuple
+   *     specifies the expected parameter type and the value to pass. The order of the arguments
+   *     must match the parameter order in the target constructor.
+   * @return A new instance of the loaded and instantiated class.
+   * @throws Exception If compilation, class loading, or instantiation fails.
+   *     <strong>Important:</strong> The target class must contain a constructor that exactly
+   *     matches the parameter types and order given in {@code args}. Otherwise, a {@code
+   *     NoSuchMethodException} will be thrown.
+   */
+  @SafeVarargs
+  public static Object loadUserInstance(
+      IPath sourcePath, String className, Tuple<Class<?>, Object>... args) throws Exception {
+
+    Class<?> newClass = DynamicCompiler.compileAndLoad(sourcePath, className);
+
+    Class<?>[] paramTypes;
+    Object[] paramValues;
+
+    if (args == null || args.length == 0) {
+      paramTypes = new Class<?>[0];
+      paramValues = new Object[0];
+    } else {
+      paramTypes = Arrays.stream(args).map(Tuple::a).toArray(Class[]::new);
+      paramValues = Arrays.stream(args).map(Tuple::b).toArray();
+    }
+
+    Constructor<?> ctor = newClass.getConstructor(paramTypes);
+    return ctor.newInstance(paramValues);
+  }
+
+  /**
+   * A custom class loader that prioritizes loading from a local compiled directory.
+   *
+   * <p>Tries to find and load a class from the given path before falling back to the system class
+   * loader. This ensures that dynamically compiled classes can override existing ones, which the
+   * standard class loader wouldn't allow.
+   */
+  private static class MyClassLoader extends URLClassLoader {
+
+    private MyClassLoader(URL[] urls) {
+      super(urls, ClassLoader.getSystemClassLoader());
+    }
+
+    @Override
+    public Class<?> loadClass(String name) throws ClassNotFoundException {
+      try {
+        return findClass(name); // Try dynamic class first
+      } catch (ClassNotFoundException e) {
+        return super.loadClass(name); // Fallback to default classpath
```

Comment:
ich habe grad darüber nachgedacht, ob wir diese hilfsklasse wirklich brauchen und nicht einfach direkt einen neuen urlclassloader anlegen und nutzen können?

wenn der eigene classloader die datei nicht findet, dann gibt es sie nicht, oder? d.h. der classloader aus der elternklasse dürfte dann auch nicht erfolgreich sein können?

Reply:
Getestet und geht nicht, dann kriegt er das nicht im internen class loader ersetzt. 

Man müsste sich tiefer mit der Implementierung des URLClassloader beschäftigen, aber das ist Grade out of scope

Reply:
hmmm. ich kann diese erfahrung empirisch bestätigen. verstehen tue ich das aber noch nicht.

Reply:
@AMatutat wir brauchen diese weiche tatsächlich. den ersten teil (`findClass`) brauchen wir für die dynamisch zu kompilierende/zu ladende klasse. und der zweite teil (`super.loadClass`) ist für das interface, was die dynamische klasse implementiert und das wir hier nicht explizit laden/kompilieren. 

habe in meinem mwe ein wenig experimentiert, und wenn das `super.loadClass` raus ist, dann findet er das gemeinsame interface nicht mehr - das ist aber wichtig für den cast.